README.md
# baucis

![Codeship](https://codeship.com/projects/43c2fac0-bfa5-0133-3ab4-4e7013906a8c/status?branch=master)

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/wprl/baucis?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![NPM](https://nodei.co/npm/baucis.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/baucis/)

Build scalable REST APIs using the open source tools and standards you and your team already know — *Mongoose, Express, and Node.js streams*.  Baucis takes the boilerplate out of building and maintaining scalable [HATEOAS](https://en.wikipedia.org/wiki/HATEOAS)/[Level 3](http://martinfowler.com/articles/richardsonMaturityModel.html) REST APIs.

Baucis has tens of thousands of users and is used in production by startups, departments, and Fortune 500 companies, and at hackathons and conferences worldwide.  MongoDB, Inc. awarded baucis first place in their 2014 Open Source Hack Contest.

Baucis is easy for beginners, easy to scale, yet flexible enough in the long term to be extended to solve complex use cases.

### Who's using Baucis?

[![Hivepod.io](http://github.com/wprl/baucis/raw/master/hivepod-io-logo.png "Hivepod.io")](https://hivepod.io) [![Kun.io App Studio LLC](http://github.com/wprl/baucis/raw/master/kunio.png "Kun.io App Studio LLC")](http://kun.io) [![Pixel Press](http://github.com/wprl/baucis/raw/master/pixel-press.jpg "Pixel Press")](http://www.projectpixelpress.com)

If you like baucis please consider adding bounties to issues in the [issue tracker](https://github.com/wprl/baucis/issues) or contributing via [GitTip](https://www.gittip.com/wprl/).


## Features

 * Automatically build controllers from your Mongoose schemata, then easily configure them.
 * Built on Express 4 so adding custom middleware is a snap.  Compatible with existing Express middleware.
 * Fully takes advantage of Node.js streaming to nimbly process large datasets.
 * Implements the HTTP specification according to the specifications ([RFC 7231](http://tools.ietf.org/rfcmarkup/7231), etc.)
 * Widely compatible with a variety of front end frameworks.
 * Perform rich queries of the API using an expressive JSON syntax via query string.
 * Supports geolocation and full text search.
 * Version your API using semver.
 * Automatically generate interactive Swagger documentation for the API.
 * Highly customizable, simple interface.  Can be extended with plugins (decorators).
 * Compatible with MongoDB 2.x and 3.x
 * Over 140 Mocha.js tests in addition to Express' and Mongoose's.


## Getting Started

To install:

    npm install --save baucis

It's easy to create an API from a Mongoose model:

``` javascript
// Create a mongoose schema.
var Vegetable = new mongoose.Schema({ name: String });
// Register new models with mongoose.
mongoose.model('vegetable', Vegetable);
// Create a simple controller.  By default these HTTP methods
// are activated: HEAD, GET, POST, PUT, DELETE
baucis.rest('vegetable');
// Create the app and listen for API requests
var app = express();
app.use('/api', baucis());
app.listen(8012);
```

That's it!  Now you have an RFC-compliant, flexible, and semantically rich API dealing with vegetables.  You could access it with URLs like `http://localhost:8012/api/vegetables`.  CRUD is supported using [GET, PUT, POST, and DELETE](https://github.com/wprl/baucis/wiki/HTTP-Verbs).


## Examples

 * [Example REST API server built with Node and Baucis](//github.com/wprl/baucis-example)
 * [Examples with Backbone.js](examples/Backbone.js)
 * [Examples with AngularJS](examples/angular-example-resource.html)
 * [Examples with Restangular](examples/angular-example-restangular.html)
 * [Examples with jQuery](examples/jQuery.js)
 * [mongoose-administration-example](https://www.npmjs.org/package/mongoose-administration-example)


## Documentation

[Check out the Wiki](https://github.com/wprl/baucis/wiki) for documentation and more in-depth information about baucis.  Check the [change log](CHANGES.md) for info on recently implemented features.


## Contact

 * Twitter [@wprl](https://twitter.com/wprl)
 * [LinkedIn](https://linkedin.com/in/willprl)
 * Email [william@kun.io](mailto:william@kun.io)

&copy; 2012-2020 Kun.io App Studio LLC



CHANGES.md
# Baucis Change Log

## v2.0.0

Update to work with latest versions of Node, Express, and Mongoose.  Special thank you to @maxiejbe

## v1.0.0-candidate.11

Baucis was awarded the MongoDB Open Source Hack Contest 2014!

Update readme and roadmap.


## v1.0.0-candidate.10

Fix some typos.


## v1.0.0-candidate.9

Add more info about `baucis.rest` to the README.


## v1.0.0-candidate.8

Use caret for dependencies.  If you're using an older npm, let's talk about a fork or other migration path.


## v1.0.0-candidate.7

README tweaks.  More tests.


## v1.0.0-candidate.6

Minor tweaks.


## v1.0.0-candidate.5

Revisit all error conditions.  Make 422 error response body an array.


## v1.0.0-candidate.4

Fix issue #183.


## v1.0.0-candidate.3

Update baucis-error version.


## v1.0.0-candidate.2

Make 200 default for empty collection/search.  Split out baucis-errors and baucis-vivify.


## v1.0.0-candidate.1

Fix #149.  Fix #181.  Move handleErrors from Api to Controller.  Add parentController method.  Add emptyCollection method to controller.  It is set to 204 by defult, but can be set to 200, 204, or 404 to set how collection endpoint requests handle no matching documents per controller.


## v1.0.0-candidate.0

Use published version of baucis-json.


## v1.0.0-prerelease.9

Allow turning off error handling per API.  Fix #178.

    baucis().handleErrors(false);


## v1.0.0-prerelease.8

Correctly handle errors in user streams.  Improve error handling.  Fix #177.


## v1.0.0-prerelease.7

Rename `baucisPath` to `fragment`.  Use a decorator for Mongoose models, rather than a wrapper object.  Wrap `mongoose.model` to add decorator to all models.  Move `lastModified` and `locking` to model objects.


## v1.0.0-prerelease.6

Move parsers/formatters from Api to baucis object.


## v1.0.0-prerelease.5

Add PATENTS file.  A little more cleaning up.


## v1.0.0-prerelease.4

Remove `request.baucis.controller`.  Remove `controller.api`.


## v1.0.0-prerelease.3

Express 4!
del -> delete
Controller activation is no longer required (again)
API activation is not required
Controller properties can change at runtime, and the API will now reflect those changes
Remove X-Powered-By for all baucis routes


## v1.0.0-prerelease.2

Merge to master and make small update to the README.


## v1.0.0-prerelease.1

Add Model.  Change X-Baucis-Update-Operator to Update-Operator.  Begin refactoring Api into smaller decorators.


## v0.20.5

Update to the latest baucis-json.  Fixes an issue where POSTing or PUTing some kinds of malformed JSON bodies could crash the instance.  Now sends a 400 Bad Request for malformed JSON.  Fix #168


## v0.20.4

Check additional opportunities to send 400 Bad Request.


## v0.20.3

README improvements.


## v0.20.2

Fix bug that overwrote Express' parent property.  Please use the `controller.children` property instead.  (Fixes #161.)


## v0.20.1

Fix publish and some typos.


## v0.20.0

Add `controller.vivify` method for embedding subcontrollers.  Deprecate `basePath` controller option.  Update configuration to "methods over values."  Move baucis errors from `baucis.errors` to `baucis.Error`.  Enable relations by default.  Improve & update README.


## v0.19.2

Handle duplicate key error as a 422.  (Fix #160).


## v0.19.1

Fix #159.


## v0.19.0

Add context to incoming and outgoing user streams. (Fix #151).


## v0.18.4

Internal refactor to use protected pipeline method for internal as well as external stream interfaces.


## v0.18.3

Update to use the latest `baucis-json`.


## v0.18.2

Don't try to set the Last-Modified header if the feature isn't enabled (fix #157).


## v0.18.1

Fix an issue where the request's content type would not be recognized when it used a MIME type parameter. (#152)


## v0.18.0

Use the mongoose `toCollectionName` util method, instead of the lingo module.


## v0.17.0

Deprecate `documents` stage of middleware.


## v0.16.21

Update dependency for deco.


## v0.16.20

Updated outdated dependencies.


## v0.16.19

Improve error messages, correct some status codes, fix issue #147.


## v0.16.18

Discriminators - correctly handle POST.  The correct model is used based on the `__t` that is passed in the request body.


## v0.16.17

Use tilde instead of caret so older npms don't fall over!


## v0.16.16

Move JSON parser and formatter to a separate module.


## v0.16.15

Fix outdated require.


## v0.16.14

Add distinct request query option.


## v0.16.13

Fix missing dependency on event-stream.


## v0.16.12

Send 415 on unparseable request content type.  Also added experimental `Api.setParser` method.


## v0.16.11

Ensure 406 is sent on requests with Accept header that cannot be fulfilled.  Fix bug with calling `api.setFormatter`.


## v0.16.10

Allow passing in map function to incoming/outgoing user streams.  It's turned into a stream internally.


## v0.16.9

Add experimental `Api.setFormatter` method.


## v0.16.8

Fix incorrect Accept header.


## v0.16.7

Allow passing in a mongoose model to the `model` controller option, in addition to string name.


## v0.16.6

Update dependency for Express to 3.5.


## v0.16.5

Fix issue #142.  Also, use event-stream over highland.


## v0.16.4

Fix issue #141


## v0.16.3

Fix issue #131


## v0.16.2

Add documentation for streaming and clean up documentation.


## v0.16.1

Republish failed `npm publish`.


## v0.16.0

Streamy goodness!  Documentation will be updated for v0.16.2.  Things are in flux...


## v0.16.0-alpha

Stream documents to manage resources more efficiently.  (Performance tests forthcoming.)

The documentation has not been updated yet... public interface subject to change.  Suggestions welcome!


## v0.15.1

Send a descriptive body for validation errors.


## v0.15.0

Send 422 on validation error.


## v0.14.2

Remove stale calls to `connect.utils.merge`.


## v0.14.1

Update to use latest `deco` features.  Other additional refactoring.


## v0.14.0


test/headers.js
var expect = require('expect.js');
var mongoose = require('mongoose');
var express = require('express');
var passport = require('passport')
var LocalStrategy = require('passport-local').Strategy;
var request = require('request').defaults({ json: true });
var baucis = require('..');

var fixtures = require('./fixtures');

describe('Headers', function () {
  before(fixtures.vegetable.init);
  beforeEach(fixtures.vegetable.create);
  after(fixtures.vegetable.deinit);

  it('sets Last-Modified for single documents', function (done) {
    var turnip = vegetables[0];
    var options = {
      url: 'http://localhost:8012/api/vegetables/' + turnip._id
    };
    request.head(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(200);
      expect(response.headers).to.have.property('last-modified');
      var modified = response.headers['last-modified'];
      var httpDate = new Date(modified).toUTCString();
      expect(modified).to.be(httpDate);

      request.get(options, function (error, response, body) {
        if (error) return done(error);
        expect(response.statusCode).to.be(200);
        expect(response.headers).to.have.property('last-modified', httpDate);
        done();
      });
    });
  });

  it('sets Last-Modified for the collection', function (done) {
    var updates = vegetables.map(function (vegetable) {
      return vegetable.lastModified;
    });
    var max = new Date(Math.max.apply(null, updates));
    var httpDate = new Date(max).toUTCString();

    var options = {
      url: 'http://localhost:8012/api/vegetables'
    };

    request.head(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(200);
      expect(response.headers).to.have.property('last-modified', httpDate);
      request.get(options, function (error, response, body) {
        if (error) return done(error);
        expect(response.statusCode).to.be(200);
        //expect(response.headers.trailer).to.be('Last-Modified, Etag');
        expect(response.headers['content-type']).to.be('application/json; charset=utf-8');
        expect(response.headers['transfer-encoding']).to.be('chunked');
        expect(response.trailers).to.have.property('last-modified', httpDate);
        done();
      });
    });
  });

  it('sets Etag for single documents', function (done) {
    var turnip = vegetables[0];
    var options = {
      url: 'http://localhost:8012/api/vegetables/' + turnip._id
    };
    request.head(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(200);
      var etag = response.headers.etag;
      expect(etag).to.match(/^"[0-9a-z]{32}"$/);
      request.get(options, function (error, response, body) {
        if (error) return done(error);
        expect(response.statusCode).to.be(200);
        expect(response.headers.etag).to.be(etag)
        done();
      });
    });
  });

  it('sets Etag for the collection', function (done) {
    var options = {
      url: 'http://localhost:8012/api/vegetables'
    };
    request.head(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(200);
      var etag = response.headers.etag;
      expect(etag).to.match(/^"[0-9a-z]{32}"$/);
      request.get(options, function (error, response, body) {
      if (error) return done(error);
        expect(response.statusCode).to.be(200);
        //expect(response.headers.trailer).to.be('Last-Modified, Etag');
        expect(response.headers['content-type']).to.be('application/json; charset=utf-8');
        expect(response.headers['transfer-encoding']).to.be('chunked');
        expect(response.trailers.etag).to.be(etag);
        done();
      });
    });
  });

  it('sets Allowed', function (done) {
    var options = {
      url: 'http://localhost:8012/api/vegetables'
    };
    request.head(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(200);
      expect(response.headers).to.have.property('allow', 'HEAD,GET,POST,PUT,DELETE');
      done();
    });
  });

  it('sends 406 Not Acceptable when the requested type is not accepted', function (done) {
    var options = {
      url: 'http://localhost:8012/api/vegetables',
      headers: {
        'Accept': 'application/xml'
      }
    };
    request.get(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(406);
      expect(response.headers).to.have.property('content-type', 'text/html; charset=utf-8');
      expect(body).to.contain('Not Acceptable: The requested content type could not be provided (406).');
      done();
    });
  });

  it('should send 415 Unsupported Media Type when the request content type cannot be parsed', function (done) {
    var options = {
      url: 'http://localhost:8012/api/vegetables',
      headers: {
        'Content-Type': 'application/xml'
      }
    };
    request.post(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(415);
      expect(body).to.have.property('message', "The request's content type is unsupported (415).");
      done();
    });
  });

  it('should match the correct MIME type, ignoring extra options and linear whitespace', function (done) {
    var options = {
      url: 'http://localhost:8012/api/vegetables',
      headers: {
        'Content-Type': '     application/json        ;       charset=UTF-8    cheese=roquefort      '
      },
      json: { name: 'Tomatillo' }
    };
    request.post(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(201);
      done();
    });
  });

  it('should not set X-Powered-By', function (done) {
    var options = {
      url: 'http://localhost:8012/api/vegetables'
    };
    request.head(options, function (error, response, body) {
      if (error) return done(error);
      expect(response.statusCode).to.be(200);
      expect(response.headers).not.to.have.property('x-powered-by');
      done();
    });
  });

});


Model.js
var deco = require('deco');
var mongoose = require('mongoose');
var pluralize = require('pluralize');

var Model = module.exports = deco(function (options, protect) {
  var model = this;

  protect.property('singular');
  protect.property('plural');
  protect.property('locking', false);
  protect.property('lastModified');

  model.deselected = function (path) {
    var deselected = [];
    // Store naming, model, and schema.
    // Find deselected paths in the schema.
    model.schema.eachPath(function (name, path) {
      if (path.options.select === false) deselected.push(name);
    });
    if (arguments.length === 0) return deselected;
    else return (deselected.indexOf(path) !== -1);
  };

  model.singular(model.modelName);
  model.plural(pluralize(model.singular()));
});

// Wrap the mongoose model function to add this mixin to all registered models.
var originalMongooseModel = mongoose.model;
mongoose.model = function () {
  var m = originalMongooseModel.apply(mongoose, arguments);
  if (!m.singular) Model.apply(m);
  return m;
};


Controller/errors.js
// __Dependencies__
var es = require('event-stream');
var util = require('util');
var mongoose = require('mongoose');
var RestError = require('rest-error');

// __Module Definition__
var decorator = module.exports = function (options, protect) {
  var baucis = require('..');
  var controller = this;
  // A controller property used to set what error status code
  // and response is sent when a query to a collection endpoint
  // yields no documents.
  protect.property('emptyCollection', 200);
  // A controller property that sets whether errors should be
  // handled if possible, or just set status code.
  protect.property('handleErrors', true, function (handle) {
    return handle ? true : false;
  });
  // If it's a Mongo bad hint error, convert to a bad request error.
  protect.use(function (error, request, response, next) {
    if (!error) return next();
    if (!error.message) return next(error);

    var message = 'The requested query hint is invalid'
    // Bad Mongo query hint (2.x).
    if (error.message === 'bad hint') {
      next(RestError.BadRequest(message));
      return;
    }
    // Bad Mongo query hint (3.x).
    if (error.message.match('planner returned error: bad hint')) {
      next(RestError.BadRequest(message));
      return;
    }
    // Bad Mongo query hint (5.x).
    if (error.message.match('planner returned error :: caused by :: hint provided does not correspond to an existing index')) {
      next(RestError.BadRequest(message));
      return;
    }
    if (!error.$err) return next(error);
    // Mongoose 3
    if (error.$err.match('planner returned error: bad hint')) {
      next(RestError.BadRequest(message));
      return;
    }
    next(error);
  });
  // Convert Mongo duplicate key error to an unprocessible entity error
  protect.use(function (error, request, response, next) {
    if (!error) return next();
    if (!error.message) return next(error);
    if (error.message.indexOf('E11000 duplicate key error') === -1) {
      next(error);
      return;
    }

    var body = {};
    var scrape = /(.*?[:]){2} (.*)[_](.*?["])(.*)(.*?["])/;
    var scraped = scrape.exec(error.message);
    var path = scraped ? scraped[2] : '???';
    var value = scraped ? scraped[4] : '???';
    body[path] = {
      message: util.format('Path `%s` (%s) must be unique.', path, value),
      originalMessage: error.message,
      name: 'MongoError',
      path: path,
      type: 'unique',
      value: value
    };

    var translatedError = RestError.UnprocessableEntity();
    translatedError.errors = body;

    next(translatedError);
  });
  // Convert Mongo validation errors to unprocessable entity errors.
  protect.use(function (error, request, response, next) {
    if (!error) return next();
    if (!(error instanceof mongoose.Error.ValidationError)) return next(error);
    var newError = RestError.UnprocessableEntity();
    newError.errors = error.errors;
    next(newError);
  });
  // Convert Mongoose version conflict error to LockConflict.
  protect.use(function (error, request, response, next) {
    if (!error) return next();
    if (!(error instanceof mongoose.Error.VersionError)) return next(error);
    next(RestError.LockConflict());
  });
  // Translate other errors to internal server errors.
  protect.use(function (error, request, response, next) {
    if (!error) return next();
    if (error instanceof RestError) return next(error);
    var error2 = RestError.InternalServerError(error.message);
    error2.stack = error.stack;
    next(error2);
  });
  // Format the error based on the Accept header.
  protect.use(function (error, request, response, next) {
    if (!error) return next();

    // Always set the status code if available.
    if (error.status >= 100) {
      response.status(error.status);
    }

    if (!controller.handleErrors()) return next(error);

    baucis.formatters(response, function (error2, formatter) {
      if (error2) return next(error2);

      var errors;

      if (!error.errors) {
        errors = [error];
      }
      else if (Array.isArray(error.errors) && error.errors.length !== 0) {
        errors = error.errors;
      }
      else {
        errors = Object.keys(error.errors).map(function (key) {
          return error.errors[key]
        });
      }

      if (errors.length === 0) {
        errors = [error];
      }

      errors = errors.map(function (error3) {
        var o = {};
        Object.getOwnPropertyNames(error3).forEach(function(key) {
          o[key] = error3[key];
        });
        delete o.domain;
        delete o.domainEmitter;
        delete o.domainThrown;
        return o;
      });

      // TODO deprecated -- always send as single error in 2.0.0
      var f = formatter(error instanceof RestError.UnprocessableEntity);
      f.on('error', next);

      es.readArray(errors).pipe(f).pipe(response);
    });
  });
};


Controller/send.js
// __Dependencies__
var es = require('event-stream');
var crypto = require('crypto');
var RestError = require('rest-error');

// __Module Definition__
var decorator = module.exports = function (options, protect) {
  var baucis = require('..');
  var controller = this;
  var lastModifiedPath = controller.model().lastModified();
  var trailers = {};

  // __Private Module Members__
  // Format the Trailer header.
  // Deprecated. See: https://stackoverflow.com/questions/22033933/using-trailer-header-with-http-chunked-transfer-how-to-set-cookie-using-it
  function addTrailer (response, header) {
    var current = response.get('Trailer');
    if (!current) response.set('Trailer', header);
    else response.set('Trailer', current + ', ' + header);
  }
  // A map that is used to create empty response body.
  function empty (context, callback) { callback(null, '') }
  // Map contexts back into documents.
  function redoc (context, callback) { callback(null, context.doc) }
  // Generate a respone Etag from a context.
  function etag (response, useTrailer) {
    if (useTrailer) {
      //addTrailer(response, 'Etag');
      response.set('Transfer-Encoding', 'chunked');
    }

    var hash = crypto.createHash('md5');

    return es.through(function (chunk) {
      hash.update(chunk);
      this.emit('data', chunk);
    },
    function () {
      if (useTrailer) {
        trailers.Etag = '"' + hash.digest('hex') + '"';
      }
      else {
        response.set('Etag', '"' + hash.digest('hex') + '"');
      }

      this.emit('end');
    });
  }

  function etagImmediate (response) {
    var hash = crypto.createHash('md5');

    return es.through(function (chunk) {
      hash.update(JSON.stringify(chunk));
      response.set('Etag', '"' + hash.digest('hex') + '"');
      this.emit('data', chunk);
    },
    function () {
      this.emit('end');
    });
  }
  // Generate a Last-Modified header/trailer
  function lastModified (response, useTrailer) {
    if (useTrailer) {
      //addTrailer(response, 'Last-Modified');
      response.set('Transfer-Encoding', 'chunked');
    }

    var latest = null;

    return es.through(function (context) {
      if (!context) return;
      if (!context.doc) return this.emit('data', context);
      if (!context.doc.get) return this.emit('data', context);

      var current = context.doc.get(lastModifiedPath);
      if (latest === null) latest = current;
      else latest = new Date(Math.max(latest, current));
      if (!useTrailer) {
        response.set('Last-Modified', latest.toUTCString());
      }
      this.emit('data', context);
    },
    function () {
      if (useTrailer) {
        if (latest) trailers['Last-Modified'] = latest.toUTCString();
      }

      this.emit('end');
    });
  }

  // Build a reduce stream.
  function reduce (accumulated, f) {
    return es.through(
      function (context) {
        accumulated = f(accumulated, context);
      },
      function () {
        this.emit('data', accumulated);
        this.emit('end');
      }
    );
  }
  // Count emissions.
  function count () {
    return reduce(0, function (a, b) { return a + 1 });
  }

  // If counting get the count and send it back directly.
  protect.finalize(function (request, response, next) {
    if (!request.baucis.count) return next();

    request.baucis.query.countDocuments(function (error, n) {
      if (error) return next(error);
      response.removeHeader('Transfer-Encoding');
      return response.json(n); // TODO support other content types
    });
  });

  // If not counting, create the basic stream pipeline.
  protect.finalize('collection', 'all',function (request, response, next) {
    var count = 0;
    var documents = request.baucis.documents;
    var pipeline = request.baucis.send = protect.pipeline(next);
    // If documents were set in the baucis hash, use them.
    if (documents) pipeline(es.readArray([].concat(documents)));
    // Otherwise, stream the relevant documents from Mongo, based on constructed query.
    else {
      pipeline(request.baucis.query.cursor());
    }
    // Map documents to contexts.
    pipeline(function (doc, callback) {
      callback(null, { doc: doc, incoming: null });
    });
    // Check for not found.
    pipeline(es.through(
      function (context) {
        count += 1;
        this.emit('data', context);
      },
      function () {
        if (count > 0) return this.emit('end');

        var status = controller.emptyCollection();
        response.status(status);

        if (status === 204) return this.emit('end');
        if (status === 200) {
          response.removeHeader('Transfer-Encoding');
          response.json([]); // TODO other content types
          this.emit('end');
          return;
        }

        this.emit('error', RestError.NotFound());
      }
    ));
    // Apply user streams.
    pipeline(request.baucis.outgoing());

    // Set the document formatter based on the Accept header of the request.
    baucis.formatters(response, function (error, formatter) {
      if (error) return next(error);
      request.baucis.formatter = formatter;
      next();
    });
  });

  protect.finalize('instance', 'all', function (request, response, next) {
    var count = 0;
    var documents = request.baucis.documents;
    var pipeline = request.baucis.send = protect.pipeline(next);
    // If documents were set in the baucis hash, use them.
    if (documents) pipeline(es.readArray([].concat(documents)));
    // Otherwise, stream the relevant documents from Mongo, based on constructed query.
    else {
      pipeline(request.baucis.query.cursor());
    }
    // Map documents to contexts.
    pipeline(function (doc, callback) {
      callback(null, { doc: doc, incoming: null });
    });
    // Check for not found.
    pipeline(es.through(
      function (context) {
        count += 1;
        this.emit('data', context);
      },
      function () {
        if (count > 0) return this.emit('end');
        this.emit('error', RestError.NotFound());
      }
    ));
    // Apply user streams.
    pipeline(request.baucis.outgoing());

    // Set the document formatter based on the Accept header of the request.
    baucis.formatters(response, function (error, formatter) {
      if (error) return next(error);
      request.baucis.formatter = formatter;
      next();
    });
  });

  // OPTIONS // TODO Express' extra handling for OPTIONS conflicts with baucis
  // TODO maybe send method names in body
  // controller.options(function (request, response, next) {
  //   console.log('here')
  //   request.baucis.send(empty);
  //   next();
  // });

  // HEAD
  protect.finalize('instance', 'head', function (request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, false));
    }

    request.baucis.send(redoc);
    request.baucis.send(etagImmediate(response));
    request.baucis.send(request.baucis.formatter());
    request.baucis.send(empty);
    next();
  });

  protect.finalize('collection', 'head', function (request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, false));
    }

    request.baucis.send(redoc);
    request.baucis.send(request.baucis.formatter(true));
    request.baucis.send(etag(response, false));
    request.baucis.send(empty);
    next();
  });

  // GET
  protect.finalize('instance', 'get', function (request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, false));
    }

    request.baucis.send(redoc);
    request.baucis.send(etagImmediate(response));
    request.baucis.send(request.baucis.formatter());
    next();
  });

  protect.finalize('collection', 'get', function (request, response, next) {
    if (lastModifiedPath) {
      request.baucis.send(lastModified(response, true));
    }

    if (request.baucis.count) {
      request.baucis.send(count());
      request.baucis.send(es.stringify());
    }
    else {
      request.baucis.send(redoc);
      request.baucis.send(request.baucis.formatter(true));
    }

    request.baucis.send(etag(response, true));
    next();
  });

  // POST
  protect.finalize('collection', 'post', function (request, response, next) {
    request.baucis.send(redoc);
    request.baucis.send(request.baucis.formatter());
    next();
  });

  // PUT
  protect.finalize('put', function (request, response, next) {
    request.baucis.send(redoc);
    request.baucis.send(request.baucis.formatter());
    next();
  });

  // DELETE
  protect.finalize('delete', function (request, response, next) {
    // Remove each document from the database.
    request.baucis.send(function (context, callback) { context.doc.remove(callback) });
    // Respond with the count of deleted documents.
    request.baucis.send(count());
    request.baucis.send(es.stringify());
    next();
  });

  protect.finalize(function (request, response, next) {
    request.baucis.send().pipe(es.through(function (chunk) {
      response.write(chunk);
    }, function () {
      response.addTrailers(trailers);
      response.end();
      this.emit('end');
    }));
  });
};
